### PR TITLE
[RTLCleanup] Merge Always block

### PR DESCRIPTION
**Pull request by prithayan**
Add support for merging always blocks in RTLCleanup.
Added just one test case for basic always block merging.
Fix for https://github.com/llvm/circt/issues/667

### CODE REVIEW

File: lib/Dialect/SV/Transforms/RTLCleanup.cpp [hunk 1]
```diff
@@ -137,10 +137,12 @@ void RTLCleanupPass::runOnGraphRegion(Region &region, bool shallow) {
     // Recursively process any regions in the op before we visit it.
     if (!shallow && op.getNumRegions() != 0)
       runOnRegionsInOp(op);
-
-    // Merge alwaysff operations by hashing them to check to see if we've
-    // already encountered one.  If so, merge them and reprocess the body.
-    if (auto alwaysOp = dyn_cast<sv::AlwaysFFOp>(op)) {
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/RTLCleanup.cpp:143]: alwaysOp is an Operation, so why not just do a ```if (isa<sv::AlwaysOp>(op) || isa<sv::AlwaysFFOp>(op))```.

File: lib/Dialect/SV/Transforms/RTLCleanup.cpp [hunk 2]
```diff
@@ -155,7 +157,9 @@ void RTLCleanupPass::runOnGraphRegion(Region &region, bool shallow) {
       // Reprocess the merged body because this may have uncovered other
       // simplifications.
       runOnGraphRegion(alwaysOp->getRegion(0), /*shallow=*/true);
-      runOnGraphRegion(alwaysOp->getRegion(1), /*shallow=*/true);
+      // AlwaysFFOp has another region, the reset block.
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/RTLCleanup.cpp:160]: could you loop over all the regions regardless of op type?

File: lib/Dialect/SV/Transforms/RTLCleanup.cpp (hunk 1 above)
Comment (prithayan) [reply lib/Dialect/SV/Transforms/RTLCleanup.cpp:143]: Done

File: lib/Dialect/SV/Transforms/RTLCleanup.cpp (hunk 2 above)
Comment (prithayan) [reply lib/Dialect/SV/Transforms/RTLCleanup.cpp]: Done

File: lib/Dialect/SV/Transforms/RTLCleanup.cpp [hunk 3]
```diff
@@ -137,10 +137,10 @@ void RTLCleanupPass::runOnGraphRegion(Region &region, bool shallow) {
...
     if (!shallow && op.getNumRegions() != 0)
       runOnRegionsInOp(op);
-
-    // Merge alwaysff operations by hashing them to check to see if we've
-    // already encountered one.  If so, merge them and reprocess the body.
-    if (auto alwaysOp = dyn_cast<sv::AlwaysFFOp>(op)) {
+    // Merge alwaysff and always operations by hashing them to check to see if
+    // we've already encountered one.  If so, merge them and reprocess the body.
+    if (isa<sv::AlwaysOp, sv::AlwaysFFOp>(op)) {
+      Operation *alwaysOp = &op;
```
Comment (darthscsi) [inline lib/Dialect/SV/Transforms/RTLCleanup.cpp:143]: Why use this temporary?

**Review by darthscsi — APPROVED** [review]
Pending unnecessary code removal.

**Comment by lattner** [comment]
@darthscsi I thought there was some concern that it wasn't safe to merge always blocks?  Isn't this why you introduced always_ff?

**Pull request merged**